Display Exploded EPubs by parsing container.opf

M0nica opened the pull request:
- Added EPub to parse .opf files and extract metadata, links, spine, resources, toc and manifestUrl for IFrameNavigator to navigate through the e-book.
- This supports the OEBPS (Open Ebook Forum Publication Structure) https://www.loc.gov/preservation/digital/formats/fdd/fdd000171.shtml 

Note: This doesn't handle loading EPubs /container.opf's from external sources. This works with locally hosted EPubs (i.e., in `static` example).

M0nica committed:
attempt to setup xml-js

M0nica committed:
update xml to json converter and tests

M0nica committed:
remove unused xml2js, use helper functions to parse xml

M0nica committed:
book is loading reader inside of reader and with broken images

M0nica committed:
add parsing for resources

M0nica committed:
treat EPub and .JSON files differently in iframenavigator

M0nica committed:
remove references to xml-js

M0nica committed:
refactor xmlToJson

M0nica committed:
undo changes to error message

M0nica committed:
update tests

M0nica commented on src/EPub.ts:
```diff
@@ -0,0 +1,244 @@
+  public constructor(manifestJSON: any, manifestUrl: URL) {
+    this.metadata = this.parseMetaData(manifestJSON);
+    //links format should be updated to point to manifest.json
+    this.links = this.parseTOC(manifestJSON);
```
```
{
        href: manifestJSON.href,
        type: "application/webpub+json",
        templated: false,
        rel: "self",
      }
``` 

`this.links` should look something like the above but it's supposed to point to a _manifest.json_ file. I am not sure if links are optional for the e-reader?

hokei replied:
Aren't the links the chapter links for the TOC?  That seems like an important thing ... but I might be mistaken

M0nica replied:
@hokei I initially thought links was another method for parsing TOC but the standard is for link to contain links to the manifest.json. That's how the other EPub in the static folder is formatting it's link (linking to manifest.json)

The spec is viewable here: https://github.com/readium/webpub-manifest#23-links 
example: 
```
"links": [
    {"rel": "self", "href": "https://example.com/manifest.json", "type": "application/webpub+json"},
    {"rel": "alternate", "href": "https://example.com/publication.epub", "type": "application/epub+zip"},
    {"rel": "search", "href": "https://example.com/search{?query}", "type": "text/html", "templated": true}
  ]
```

M0nica commented on test/EPubTests.ts:
```diff
@@ -0,0 +1,452 @@
+      expect(manifest.metadata.title).to.equal("The Elephant");
+    });
+
+    it.skip("should store links", () => {
```
TBD on if links is necessary for the EPub.ts files

M0nica committed:
update mock data

M0nica commented on test/EPubTests.ts:
```diff
@@ -0,0 +1,452 @@
+      });
+    });
+
+    it.skip("should return the response from fetch, and save it to local store", async () => {
```
I modified this test suite from Manifest.ts but need to understand it better. I am not sure why xx is returning the manifest but saying that an empty response was expected 🤔 


AssertionError: expected { Object (metadata, links, ...) } to deeply equal { Object (metadata, links, ...) }
  ```
      + expected - actual

       {
         "links": []
         "manifestUrl": {}
      -  "metadata": {
      -    "title": "The Elephant"
      -  }
      +  "metadata": {}
         "resources": []
         "spine": []
         "toc": []
       }
```

 I am not sure why it's expecting an empty object, I am expecting ..

If I switch expected to  hard-coded 
```{
          links: [],
          manifestUrl: {},
          metadata: {
            title: "The Elephant",
          },
          resources: [],
          spine: [],
          toc: [],
        }
```

 it fails with `AssertionError: expected { Object (metadata, links, ...) } to deeply equal {}

M0nica commented on src/EPub.ts:
```diff
@@ -0,0 +1,244 @@
+  public parseMetaData(manifestJSON: any): any {
+    return manifestJSON?.package?.metadata
+      ? {
+          title: manifestJSON.package.metadata["dc:title"]["#text"],
```
This metadata assignment needs to be updated to dynamically parse more robust metadata from the package.opf:  
```

dc:creator | <empty> or <unknown> | author
-- | -- | --
dc:contributor | <empty> or <unknown> | contributor
dc:creator or dc:contributor | aut | author
dc:creator or dc:contributor | pbl | publisher
dc:creator or dc:contributor | trl | translator
dc:creator or dc:contributor | edt | editor
dc:creator or dc:contributor | ill | illustrator
dc:creator or dc:contributor | art | artist
dc:creator or dc:contributor | clr | colorist
dc:creator or dc:contributor | nrt | narrator
dc:publisher | N/A | publisher
media:narrator | N/A | narrator
```
source: https://github.com/readium/architecture/blob/master/streamer/parser/metadata.md 

Does the e-reader only care about a subset of potential metadata?

hokei replied:
It should parse all of the metadata, if available.  We shouldn't strip it out ...

hokei replied:
Unless this is used to explicitly populate a title or something ... (Where is it used?)

hokei replied:
It looks like the only place that `Manifest.metadata` is used is to set the title in `IFrameNavigator.ts#L899`

In that case you might want to replace `Manifest.metadata` with `Manifest.title` for clarity and save some work

hokei commented on src/EPub.ts:
```diff
@@ -0,0 +1,244 @@
+  [key: string]: string[] | string | xmlObject | [];
+};
+
+function xmlToJson(xml: any) {
```
does DomParser not work here? (https://developer.mozilla.org/en-US/docs/Web/API/DOMParser)

M0nica replied:
This function is transforming the output from DomParser into JSON. 

 https://github.com/NYPL-Simplified/webpub-viewer/blob/6a4325c2ee3c836e0a9af413ef1f25640121e8e3/src/Manifest.ts#L85-L88

It looks like the MIME types that are supported by DomParser are: 

text/html | Document
-- | --
text/xml | XMLDocument
application/xml | XMLDocument
application/xhtml+xml | XMLDocument
image/svg+xml | XMLDocument

hokei commented on src/EPub.ts:
```diff
@@ -0,0 +1,244 @@
+  children?: Array<Link>;
+}
+
+export default class Manifest {
```
Rename classname so it doesn't clash with `Manifest.ts`

hokei commented on src/EPub.ts:
```diff
@@ -0,0 +1,244 @@
+      return new Manifest(JSON.parse(manifestJSON), manifestUrl);
+    };
+
+    const tryToUpdateManifestButIgnoreResult = async (): Promise<void> => {
```
Some of these methods are exactly the same as `Manifest.ts`
I wonder if there's a way to either rework `Manifest.ts` to recognize if it's an epub or not, or extract similar functions, or find some other way to consolidate this code so we don't have two copies of very similar code.  It'll also make it more obvious which parts of this are specific to the webpub stuff and which are specific to epub.

M0nica replied:
I updated to have all of the code in one file for parsing manifest.json and package.opf in the same file. I wonder if it'll still be confusing to refer to both formats as Manifest

hokei replied:
yeah definitely, it's not really "manifest" as it is some kind of ... parser for the manifest-type file?  Aaaah names are hard.

M0nica replied:
I renamed the parsed `package.opf` to `OPFPackage` in certain places for now.

M0nica committed:
consolidate EPub + Manifest

M0nica committed:
fix types for helper functions

M0nica committed:
parse JSON before extracting data

M0nica committed:
fix issue with spine href values being undefined

M0nica committed:
update manifest parsing

M0nica committed:
update streamed server to check if .js file exists

M0nica committed:
resolve failing manifest.ts tests

M0nica committed:
remove unused commented out console logs

M0nica committed:
rename ManifestJSON to OPFPackage in certain places

M0nica commented on examples/streamed/streamed-server.js:
```diff
@@ -1,54 +1,68 @@
-    this.expressApp.use("/readerNYPL", express.static(__dirname + "/../../viewer"));
+    this.expressApp.use(
+      "/readerNYPL",
+      express.static(__dirname + "/../../viewer", { extensions: ["js"] })
```
This change here makes is to so the express will look for missing files with the `.js` extension so if it’s trying to get
http://localhost:4444/readerNYPL/HTMLUtilities and it 404s
it should check for `
http://localhost:4444/readerNYPL/HTMLUtilities.js` and only return a 404 if that is not found

hokei commented on examples/streamed/streamed-server.js:
```diff
@@ -1,54 +1,68 @@
-var express = require('express');
+var express = require("express");
```
if it's not much trouble can you commit your vscode settings as well so we can stay in sync?  I'm not sure I'm using anything, as evidenced by the inconsistent nature of this file.

M0nica replied:
I added `.prettierrc.json` with my configuration

hokei commented:
can you add everything under `/viewer` to gitignore?  I don't think there's a reason it should be committed ...

M0nica committed:
update gitignore to ignore all viewer files except index.html

M0nica committed:
remove generated viewer files from git

M0nica committed:
add prettier config

M0nica commented:
> can you add everything under `/viewer` to gitignore? I don't think there's a reason it should be committed ...

I updated to add all of the files in `/viewer` aside from index.html to the .gitignore and removed all of the `/viewer` files from git.

M0nica committed:
update to not ignore fetch and require!

M0nica committed:
make parsing of EPub more robust  and add try catch logic

M0nica committed:
move OPF declaration closer to usage and run all tests

hokei commented on src/Manifest.ts:
```diff
@@ -1,135 +1,284 @@
+    const fetchManifest = async (): Promise<Manifest> => {
+      const isJSONManifest = Boolean(manifestUrl.href.endsWith(".json"));
+
+      const manifestJSON = isJSONManifest
```
this could use some comments to make sure we know what is happening?  Is this method only for json manifests?

hokei commented on src/Manifest.ts:
```diff
@@ -1,135 +1,284 @@
+  children?: Array<Link>;
 }
 
 export default class Manifest {
```
Can you put a comment here on what exactly a Manifest is in this context (making it clear that it can be a Package Document or a Webpub Manifest?)

hokei commented on src/Manifest.ts:
```diff
@@ -1,135 +1,284 @@
+    return OPF;
+  }
+
+  public parseMetaData(OPFPackage: any): any {
```
Can we have a block comment making clear that there's a whole block of OPF-specific functions?

hokei reviewed (approved):
Looks good, I think it would be helpful to have some comments to help us parse this in the future.

M0nica committed:
add clarifying comments and variable names

M0nica committed:
bump version to 0.1.2
